### PR TITLE
Make `E0117` error clear

### LINKED ISSUE #96227: Make "only traits defined in the current crate can be implemented for arbitrary types" less vague
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b04540728eae88b27ee5d19b5bc0b685

```rust
use chrono::offset::Utc;
use std::fmt::Display;

impl Display for Utc {}
fn main() {}
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[[E0117]](https://doc.rust-lang.org/stable/error-index.html#E0117): only traits defined in the current crate can be implemented for arbitrary types
 [--> src/main.rs:4:1
](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3ba9fdd15e921f6ac1ecff15c7d0b783#)  |
4 | impl Display for Utc {}
  | ^^^^^^^^^^^^^^^^^---
  | |                |
  | |                `Utc` is not defined in the current crate
  | impl doesn't use only types from inside the current crate
  |
  = note: define and implement a trait or new type instead

For more information about this error, try `rustc --explain E0117`.
error: could not compile `playground` due to previous error
```

Ideally the output should look like:

Not sure... my main issue with this error is the message itself: "only traits defined in the current crate can be implemented for arbitrary types."  The fact that it only mentioned "arbitrary types" leads me to believe that my only solution is to change something about the type, but I don't what that change Is supposed to be (In this case, it is using the ["newtype" idiom](https://doc.rust-lang.org/rust-by-example/generics/new_types.html). 

The word `arbitrary` is vague to me - as a beginner, I don't fully know what that means.  In fact, if I had to guess, I would consider both the original type, and the new type that wraps the old type, both "arbitrary," so I think the issue is around that word.  I feel as if the message should not mention "arbitrary" and should rather indicate that you can only implement this trait on a type that's defined within this crate.

I do understand that the help diagnostic is nudging the user towards a newtype pattern, but my main point is that the error is pretty ambiguous to me.

### PR DESCRIPTION
**Pull request by TaKO8Ki**
closes #96227

### CODE REVIEW

File: compiler/rustc_typeck/src/coherence/orphan.rs [hunk 1]
```diff
@@ -201,18 +202,24 @@ fn emit_orphan_check_error<'tcx>(
...
+    self_ty: Ty<'tcx>,
     self_ty_span: Span,
     generics: &hir::Generics<'tcx>,
     err: traits::OrphanCheckErr<'tcx>,
 ) -> Result<!, ErrorGuaranteed> {
     Err(match err {
         traits::OrphanCheckErr::NonLocalInputType(tys) => {
+            let msg = match self_ty.kind() {
+                ty::Adt(..) => "can be implemented for types defined outside of the crate",
+                ty::Param(_) => "can have blanket implementations defined in this trait",
```
Comment (davidtwco) [inline compiler/rustc_typeck/src/coherence/orphan.rs:214]: Can you add a test case that emits this?

**Review by davidtwco — CHANGES_REQUESTED** [review]
Looks good to me, left a comment.

File: compiler/rustc_typeck/src/coherence/orphan.rs (hunk 1 above)
Comment (TaKO8Ki) [reply compiler/rustc_typeck/src/coherence/orphan.rs:214]: The following lines seem to correspond to errors for `ty::Param`, so I'll remove this line.

https://github.com/rust-lang/rust/blob/10baaa6ed243c3e026491b1068c44356eaf24069/compiler/rustc_typeck/src/coherence/orphan.rs#L311-L314

**Review by davidtwco — APPROVED** [review]

**Pull request merged**